phoenix-pherf/src/main/java/org/apache/phoenix/pherf/jmx/MonitorManager.java

```diff
@@ -0,0 +1,172 @@
+    // List of MonitorDetails for all the running monitors.
+    // TODO Move this out to config. Possible use Guice and use IOC to inject it in.
+    private static final List<MonitorDetails> MONITOR_DETAILS_LIST =
+            Arrays.asList(MonitorDetails.values());
+    private final ResultHandler resultHandler;
+    private final long monitorFrequency;
+    private volatile long rowCount;
+    private volatile boolean shouldStop = false;
+    private volatile boolean isRunning = false;
+
+    public MonitorManager() throws Exception {
+        this(PherfConstants.MONITOR_FREQUENCY);
+    }
+
+    /**
+     *
+     * @param monitorFrequency Frequency at which monitor stats are written to a log file.
+     * @throws Exception
+     */
+    public MonitorManager(long monitorFrequency) throws Exception {
+        this.monitorFrequency = monitorFrequency;
+        MBeanServer mbs = ManagementFactory.getPlatformMBeanServer();
+
+        // Register all the monitors to JMX
+        for (MonitorDetails monitorDetails : MONITOR_DETAILS_LIST) {
+            StandardMBean bean = new StandardMBean(monitorDetails.getMonitor(), Monitor.class);
+            ObjectName monitorThreadStatName = new ObjectName(monitorDetails.toString());
+            try {
+                mbs.registerMBean(bean, monitorThreadStatName);
+            } catch (InstanceAlreadyExistsException e) {
+                mbs.unregisterMBean(monitorThreadStatName);
+                mbs.registerMBean(bean, monitorThreadStatName);
+            }
+        }
+        this.rowCount = 0;
+        this.resultHandler = new CSVResultHandler(PherfConstants.MONITOR_FILE_NAME, ResultFileDetails.CSV);
+    }
+
+    @Override
+    public void run() {
+        try {
+            while (!shouldStop()) {
+                isRunning = true;
+                List rowValues = new ArrayList<String>();
+                synchronized (resultHandler) {
+                    for (MonitorDetails monitorDetails : MONITOR_DETAILS_LIST) {
+                        rowValues.clear();
+                        try {
+                            StandardMBean bean = new StandardMBean(monitorDetails.getMonitor(), Monitor.class);
+
+                            Calendar calendar = new GregorianCalendar();
+                            rowValues.add(monitorDetails);
+
+                            rowValues.add(((Monitor) bean.getImplementation()).getStat());
+                            rowValues.add(DateUtil.DEFAULT_MS_DATE_FORMATTER.format(calendar.getTime()));
+                            Result
+                                    result = new Result(ResultFileDetails.CSV, ResultFileDetails.CSV_MONITOR.getHeader().toString(), rowValues);
+                            resultHandler.write(result);
+                        } catch (Exception e) {
+                            throw new FileLoaderRuntimeException("Could not log monitor result.", e);
+                        }
+                        rowCount++;
```

Comment:
The synchronization is broken because reading rowCount in getRowCount() is done by locking the "this" object and in run() modifying rowCount is guarded by locking resultHandler member variable. If you have two threads, one thread (thread 1) executing run() and other thread (thread2) executing getRowCount() concurrently, then thread2 won't wait for thread1 to finish or vice-versa. As a result you will have two threads reading and writing rowCount at the same time. If you don't care about consistent row count, then remove synchronized from getRowCount altogether. If you do care, then making rowCount an AtomicLong is advisable.